Revise when trowing AbortedException, PrematureCloseException and InternalNettyException

Related to https://github.com/spring-projects/spring-framework/issues/24596

https://github.com/reactor/reactor-netty/commit/5059f86a8e3c6e00d0859d56f869f704019cef61#diff-03cecb2a278ad2453fcf6a85575dae60

Hello @violetagg, any news about this issue? We've been experiencing a problem similar to the one described in https://github.com/spring-projects/spring-framework/issues/24596.

We are using Spring Boot Webflux 2.2.5.RELEASE and reactor-netty 0.9.6.RELEASE. Sometimes our client is closing connection because of timeout and we're getting something like this in our logs:

```
[e26eee7f-8975910] Error [java.lang.UnsupportedOperationException] for HTTP GET "/status/ping", but ServerHttpResponse already committed (200 OK)
```
and:
```
reactor.netty.ReactorNetty$InternalNettyException: java.nio.channels.ClosedChannelException
	Suppressed: reactor.core.publisher.FluxOnAssembly$OnAssemblyException: 
Error has been observed at the following site(s):
	|_ checkpoint ⇢ Handler org.springframework.boot.actuate.endpoint.web.reactive.AbstractWebFluxEndpointHandlerMapping$ReadOperationHandler#handle(ServerWebExchange) [DispatcherHandler]
Stack trace:
Caused by: java.nio.channels.ClosedChannelException: null
	at io.netty.channel.AbstractChannel$AbstractUnsafe.newClosedChannelException(AbstractChannel.java:957)
	at io.netty.channel.AbstractChannel$AbstractUnsafe.write(AbstractChannel.java:865)
	at io.netty.channel.DefaultChannelPipeline$HeadContext.write(DefaultChannelPipeline.java:1367)
	at io.netty.channel.AbstractChannelHandlerContext.invokeWrite0(AbstractChannelHandlerContext.java:717)
	at io.netty.channel.AbstractChannelHandlerContext.invokeWrite(AbstractChannelHandlerContext.java:709)
	at io.netty.channel.AbstractChannelHandlerContext.write(AbstractChannelHandlerContext.java:792)
	at io.netty.channel.AbstractChannelHandlerContext.write(AbstractChannelHandlerContext.java:702)
	at io.netty.channel.CombinedChannelDuplexHandler$DelegatingChannelHandlerContext.write(CombinedChannelDuplexHandler.java:526)
	at io.netty.handler.codec.MessageToMessageEncoder.write(MessageToMessageEncoder.java:112)
	at io.netty.channel.CombinedChannelDuplexHandler.write(CombinedChannelDuplexHandler.java:346)
	at io.netty.channel.AbstractChannelHandlerContext.invokeWrite0(AbstractChannelHandlerContext.java:717)
	at io.netty.channel.AbstractChannelHandlerContext.invokeWrite(AbstractChannelHandlerContext.java:709)
	at io.netty.channel.AbstractChannelHandlerContext.write(AbstractChannelHandlerContext.java:792)
	at io.netty.channel.AbstractChannelHandlerContext.write(AbstractChannelHandlerContext.java:702)
	at reactor.netty.http.server.HttpTrafficHandler.write(HttpTrafficHandler.java:269)
	at io.netty.channel.AbstractChannelHandlerContext.invokeWrite0(AbstractChannelHandlerContext.java:717)
	at io.netty.channel.AbstractChannelHandlerContext.invokeWriteAndFlush(AbstractChannelHandlerContext.java:764)
	at io.netty.channel.AbstractChannelHandlerContext$WriteTask.run(AbstractChannelHandlerContext.java:1104)
	at io.netty.util.concurrent.AbstractEventExecutor.safeExecute(AbstractEventExecutor.java:164)
	at io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:472)
	at io.netty.channel.epoll.EpollEventLoop.run(EpollEventLoop.java:384)
	at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:989)
	at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadL
```

--- a/src/main/java/reactor/netty/FutureMono.java
+++ b/src/main/java/reactor/netty/FutureMono.java
@@ -29,6 +29,7 @@ import org.reactivestreams.Subscription;
 import reactor.core.CoreSubscriber;
 import reactor.core.publisher.Mono;
 import reactor.core.publisher.Operators;
+import reactor.netty.channel.AbortedException;
 import reactor.util.context.Context;
 
 /**
@@ -263,7 +264,7 @@ public abstract class FutureMono extends Mono<Void> {
 		@Override
 		@SuppressWarnings("unchecked")
 		public void operationComplete(F future) {
-			if (!future.isSuccess()) {//Avoid singleton
+			if (!future.isSuccess()) {
 				s.onError(wrapError(future.cause()));
 			}
 			else {
@@ -272,10 +273,10 @@ public abstract class FutureMono extends Mono<Void> {
 		}
 
 		private static Throwable wrapError(Throwable error) {
-			if(error instanceof ClosedChannelException) {
-				//Update with a common aborted exception?
-				return ReactorNetty.wrapException(error);
-			} else {
+			if (error instanceof ClosedChannelException) {
+				return new AbortedException(error);
+			}
+			else {
 				return error;
 			}
 		}

--- a/src/main/java/reactor/netty/channel/AbortedException.java
+++ b/src/main/java/reactor/netty/channel/AbortedException.java
@@ -26,9 +26,14 @@ import java.net.SocketException;
  * @since 0.6
  */
 public class AbortedException extends RuntimeException {
+	static final String CONNECTION_CLOSED_BEFORE_SEND = "Connection has been closed BEFORE send operation";
 
 	public AbortedException(String message) {
 		super(message);
+	}
+
+	public AbortedException(Throwable throwable) {
+		super(throwable);
 	}
 
 	/**
@@ -40,7 +45,7 @@ public class AbortedException extends RuntimeException {
 	 * @return true if connection has been simply aborted on a tcp level
 	 */
 	public static boolean isConnectionReset(Throwable err) {
-		return err instanceof AbortedException ||
+		return (err instanceof AbortedException && CONNECTION_CLOSED_BEFORE_SEND.equals(err.getMessage())) ||
 		       (err instanceof IOException && (err.getMessage() == null ||
 		                                       err.getMessage()
 		                                          .contains("Broken pipe") ||
@@ -49,5 +54,9 @@ public class AbortedException extends RuntimeException {
 		       (err instanceof SocketException && err.getMessage() != null &&
 		                                          err.getMessage()
 		                                             .contains("Connection reset by peer"));
+	}
+
+	public static AbortedException beforeSend() {
+		return new AbortedException(CONNECTION_CLOSED_BEFORE_SEND);
 	}
 }

--- a/src/main/java/reactor/netty/channel/ChannelOperations.java
+++ b/src/main/java/reactor/netty/channel/ChannelOperations.java
@@ -16,6 +16,7 @@
 
 package reactor.netty.channel;
 
+import java.nio.channels.ClosedChannelException;
 import java.util.Objects;
 import java.util.concurrent.Callable;
 import java.util.concurrent.atomic.AtomicReferenceFieldUpdater;
@@ -241,7 +242,7 @@ public class ChannelOperations<INBOUND extends NettyInbound, OUTBOUND extends Ne
 	@SuppressWarnings("unchecked")
 	public NettyOutbound send(Publisher<? extends ByteBuf> dataStream) {
 		if (!channel().isActive()) {
-			return then(Mono.error(new AbortedException("Connection has been closed BEFORE send operation")));
+			return then(Mono.error(AbortedException.beforeSend()));
 		}
 		if (dataStream instanceof Mono) {
 			return then(((Mono<?>)dataStream).flatMap(m -> FutureMono.from(channel().writeAndFlush(m)))
@@ -254,7 +255,7 @@ public class ChannelOperations<INBOUND extends NettyInbound, OUTBOUND extends Ne
 	@SuppressWarnings("unchecked")
 	public NettyOutbound sendObject(Publisher<?> dataStream) {
 		if (!channel().isActive()) {
-			return then(Mono.error(new AbortedException("Connection has been closed BEFORE send operation")));
+			return then(Mono.error(AbortedException.beforeSend()));
 		}
 		if (dataStream instanceof Mono) {
 			return then(((Mono<?>)dataStream).flatMap(m -> FutureMono.from(channel().writeAndFlush(m)))
@@ -267,7 +268,7 @@ public class ChannelOperations<INBOUND extends NettyInbound, OUTBOUND extends Ne
 	public NettyOutbound sendObject(Object message) {
 		if (!channel().isActive()) {
 			ReactorNetty.safeRelease(message);
-			return then(Mono.error(new AbortedException("Connection has been closed BEFORE send operation")));
+			return then(Mono.error(AbortedException.beforeSend()));
 		}
 		return then(FutureMono.deferFuture(() -> connection.channel()
 		                                                   .writeAndFlush(message)),
@@ -465,6 +466,23 @@ public class ChannelOperations<INBOUND extends NettyInbound, OUTBOUND extends Ne
 	protected final String formatName() {
 		return getClass().getSimpleName()
 		                 .replace("Operations", "");
+	}
+
+	/**
+	 * Wrap an inbound error
+	 *
+	 * @param err the {@link Throwable} cause
+	 */
+	protected Throwable wrapInboundError(Throwable err) {
+		if (err instanceof ClosedChannelException) {
+			return new AbortedException(err);
+		}
+		else if (err instanceof OutOfMemoryError) {
+			return ReactorNetty.wrapException(err);
+		}
+		else {
+			return err;
+		}
 	}
 
 	@Override

--- a/src/main/java/reactor/netty/channel/FluxReceive.java
+++ b/src/main/java/reactor/netty/channel/FluxReceive.java
@@ -32,7 +32,6 @@ import reactor.core.CoreSubscriber;
 import reactor.core.Disposable;
 import reactor.core.publisher.Flux;
 import reactor.core.publisher.Operators;
-import reactor.netty.ReactorNetty;
 import reactor.util.Logger;
 import reactor.util.Loggers;
 import reactor.util.concurrent.Queues;
@@ -385,7 +384,7 @@ final class FluxReceive extends Flux<Object> implements Subscription, Disposable
 		}
 
 		if (err instanceof OutOfMemoryError) {
-			this.inboundError = ReactorNetty.wrapException(err);
+			this.inboundError = parent.wrapInboundError(err);
 			try {
 				if (receiver != null) {
 					// propagate java.lang.OutOfMemoryError: Direct buffer memory
@@ -399,14 +398,14 @@ final class FluxReceive extends Flux<Object> implements Subscription, Disposable
 			}
 		}
 		else if (err instanceof ClosedChannelException) {
-			this.inboundError = ReactorNetty.wrapException(err);
+			this.inboundError = parent.wrapInboundError(err);
 		}
 		else {
 			this.inboundError = err;
 		}
 
 		if (receiverFastpath && receiver != null) {
-			receiver.onError(err);
+			receiver.onError(inboundError);
 		}
 		else {
 			drainReceiver();

--- a/src/main/java/reactor/netty/channel/MonoSendMany.java
+++ b/src/main/java/reactor/netty/channel/MonoSendMany.java
@@ -161,7 +161,7 @@ final class MonoSendMany<I, O> extends MonoSend<I, O> implements Scannable {
 			}
 
 			if (t instanceof ClosedChannelException) {
-				t = ReactorNetty.wrapException(t);
+				t = new AbortedException(t);
 			}
 
 			terminalSignal = t;

--- a/src/main/java/reactor/netty/http/HttpOperations.java
+++ b/src/main/java/reactor/netty/http/HttpOperations.java
@@ -100,7 +100,7 @@ public abstract class HttpOperations<INBOUND extends NettyInbound, OUTBOUND exte
 	@SuppressWarnings("unchecked")
 	public NettyOutbound send(Publisher<? extends ByteBuf> source) {
 		if (!channel().isActive()) {
-			return then(Mono.error(new AbortedException("Connection has been closed BEFORE send operation")));
+			return then(Mono.error(AbortedException.beforeSend()));
 		}
 		if (source instanceof Mono) {
 			return new PostHeadersNettyOutbound(((Mono<ByteBuf>)source)
@@ -132,7 +132,7 @@ public abstract class HttpOperations<INBOUND extends NettyInbound, OUTBOUND exte
 	public NettyOutbound sendObject(Object message) {
 		if (!channel().isActive()) {
 			ReactorNetty.safeRelease(message);
-			return then(Mono.error(new AbortedException("Connection has been closed BEFORE send operation")));
+			return then(Mono.error(AbortedException.beforeSend()));
 		}
 		if (!(message instanceof ByteBuf)) {
 			return super.sendObject(message);
@@ -162,7 +162,7 @@ public abstract class HttpOperations<INBOUND extends NettyInbound, OUTBOUND exte
 	@Override
 	public Mono<Void> then() {
 		if (!channel().isActive()) {
-			return Mono.error(new AbortedException("Connection has been closed BEFORE send operation"));
+			return Mono.error(AbortedException.beforeSend());
 		}
 
 		if (hasSentHeaders()) {

--- a/src/main/java/reactor/netty/http/client/HttpClientOperations.java
+++ b/src/main/java/reactor/netty/http/client/HttpClientOperations.java
@@ -19,6 +19,7 @@ package reactor.netty.http.client;
 import java.net.InetSocketAddress;
 import java.net.URI;
 import java.net.URISyntaxException;
+import java.nio.channels.ClosedChannelException;
 import java.util.Collections;
 import java.util.Map;
 import java.util.Objects;
@@ -352,7 +353,7 @@ class HttpClientOperations extends HttpOperations<NettyInbound, NettyOutbound>
 	@Override
 	public NettyOutbound send(Publisher<? extends ByteBuf> source) {
 		if (!channel().isActive()) {
-			return then(Mono.error(new AbortedException("Connection has been closed BEFORE send operation")));
+			return then(Mono.error(AbortedException.beforeSend()));
 		}
 		if (source instanceof Mono) {
 			return super.send(source);
@@ -650,13 +651,21 @@ class HttpClientOperations extends HttpOperations<NettyInbound, NettyOutbound>
 		return request;
 	}
 
+	@Override
+	protected Throwable wrapInboundError(Throwable err) {
+		if (err instanceof ClosedChannelException) {
+			return new PrematureCloseException(err);
+		}
+		return super.wrapInboundError(err);
+	}
+
 	final HttpRequest getNettyRequest() {
 		return nettyRequest;
 	}
 
 	final Mono<Void> send() {
 		if (!channel().isActive()) {
-			return Mono.error(new AbortedException("Connection has been closed BEFORE send operation"));
+			return Mono.error(AbortedException.beforeSend());
 		}
 		if (markSentHeaderAndBody()) {
 			HttpMessage request = newFullBodyMessage(Unpooled.EMPTY_BUFFER);

--- a/src/main/java/reactor/netty/http/client/PrematureCloseException.java
+++ b/src/main/java/reactor/netty/http/client/PrematureCloseException.java
@@ -16,10 +16,16 @@
 
 package reactor.netty.http.client;
 
+import io.netty.channel.ChannelInboundHandler;
+import reactor.netty.channel.ChannelOperations;
+
 import java.io.IOException;
 
 /**
  * An error for signalling that the connection was closed prematurely
+ * {@link ChannelInboundHandler#channelInactive(io.netty.channel.ChannelHandlerContext)},
+ * {@link ChannelOperations#onInboundClose()},
+ * {@link ChannelOperations#onInboundError()}
  *
  * @author Violeta Georgieva
  */
@@ -48,6 +54,10 @@ public final class PrematureCloseException extends IOException {
 
 	PrematureCloseException(String message) {
 		super(message);
+	}
+
+	PrematureCloseException(Throwable throwable) {
+		super(throwable);
 	}
 
 	@Override

--- a/src/test/java/reactor/netty/FutureMonoTest.java
+++ b/src/test/java/reactor/netty/FutureMonoTest.java
@@ -17,6 +17,7 @@ package reactor.netty;
 
 import io.netty.util.concurrent.*;
 import org.junit.Test;
+import reactor.netty.channel.AbortedException;
 import reactor.test.StepVerifier;
 
 import java.nio.channels.ClosedChannelException;
@@ -31,7 +32,7 @@ public class FutureMonoTest {
 		Future<Void> promise = eventExecutor.newFailedFuture(new ClosedChannelException());
 
 		StepVerifier.create(FutureMono.from(promise))
-		            .expectError(ReactorNetty.InternalNettyException.class)
+		            .expectError(AbortedException.class)
 		            .verify(Duration.ofSeconds(30));
 	}
 
@@ -45,7 +46,7 @@ public class FutureMonoTest {
 		StepVerifier.create(FutureMono.from(promise))
 		            .expectSubscription()
 		            .then(() -> promise.setFailure(new ClosedChannelException()))
-		            .expectError(ReactorNetty.InternalNettyException.class)
+		            .expectError(AbortedException.class)
 		            .verify(Duration.ofSeconds(30));
 	}
 
@@ -55,7 +56,7 @@ public class FutureMonoTest {
 		Supplier<Future<Void>> promiseSupplier = () -> eventExecutor.newFailedFuture(new ClosedChannelException());
 
 		StepVerifier.create(FutureMono.deferFuture(promiseSupplier))
-		            .expectError(ReactorNetty.InternalNettyException.class)
+		            .expectError(AbortedException.class)
 		            .verify(Duration.ofSeconds(30));
 	}
 
@@ -70,7 +71,7 @@ public class FutureMonoTest {
 		StepVerifier.create(FutureMono.deferFuture(promiseSupplier))
 		            .expectSubscription()
 		            .then(() -> promise.setFailure(new ClosedChannelException()))
-		            .expectError(ReactorNetty.InternalNettyException.class)
+		            .expectError(AbortedException.class)
 		            .verify(Duration.ofSeconds(30));
 	}
 }

